pkg/testutils/fingerprintutils/fingerprint.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package fingerprintutils
+
+import (
+	"context"
+	gosql "database/sql"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/gostdlib/x/sync/errgroup"
+)
+
+type FingerprintOption struct {
+	AOST            hlc.Timestamp
+	Stripped        bool
+	RevisionHistory bool
+	StartTime       hlc.Timestamp
+}
+
+func AOST(aost hlc.Timestamp) func(*FingerprintOption) {
+	return func(opt *FingerprintOption) {
+		opt.AOST = aost
+	}
+}
+
+func Stripped() func(*FingerprintOption) {
+	return func(opt *FingerprintOption) {
+		opt.Stripped = true
+	}
+}
+
+func RevisionHistory() func(*FingerprintOption) {
+	return func(opt *FingerprintOption) {
+		opt.RevisionHistory = true
+	}
+}
+
+func StartTime(startTime hlc.Timestamp) func(*FingerprintOption) {
+	return func(opt *FingerprintOption) {
+		opt.StartTime = startTime
+	}
+}
+
+func getOpts(optFuncs ...func(*FingerprintOption)) (*FingerprintOption, error) {
+	fingerprintOpts := &FingerprintOption{}
+	for _, opt := range optFuncs {
+		opt(fingerprintOpts)
+	}
+
+	if fingerprintOpts.Stripped && !fingerprintOpts.StartTime.IsEmpty() {
+		return nil, errors.New("cannot specify stripped and a start time")
+	}
+	if fingerprintOpts.Stripped && fingerprintOpts.RevisionHistory {
+		return nil, errors.New("cannot specify stripped and revision history")
+	}
+	return fingerprintOpts, nil
+}
+
+func FingerprintTable(
+	ctx context.Context, db *gosql.DB, tableID uint32, optFuncs ...func(*FingerprintOption),
+) (int64, error) {
+	opts, err := getOpts(optFuncs...)
+	if err != nil {
+		return 0, err
+	}
+
+	aostCmd := ""
+	if !opts.AOST.IsEmpty() {
+		aostCmd = fmt.Sprintf("AS OF SYSTEM TIME '%s'", opts.AOST.AsOfSystemTime())
+	}
+
+	cmd := fmt.Sprintf(`SELECT * FROM crdb_internal.fingerprint(crdb_internal.table_span(%d),true) %s`, tableID, aostCmd)
+
+	if !opts.Stripped {
+		startTime := opts.StartTime.AsOfSystemTime()
+		cmd = fmt.Sprintf(`SELECT * FROM crdb_internal.fingerprint(crdb_internal.table_span(%d),'%s'::DECIMAL,%t) %s`, tableID, startTime, opts.RevisionHistory, aostCmd)
+	}
+	var fingerprint int64
+	row := db.QueryRowContext(ctx, cmd)
+	if err := row.Scan(&fingerprint); err != nil {
+		return 0, fmt.Errorf("fingerprint command failed on table id %d: %w", tableID, err)
```

Comment:
This is fine, but here and elsewhere we may consider `errors.Wrapf`.

Reply:
Done. subbed it in everywhere.